README.md
qiitr - R Binding for Qiita API
==========================

[![R build status](https://github.com/yutannihilation/qiitr/workflows/R-CMD-check/badge.svg)](https://github.com/yutannihilation/qiitr/actions)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/qiitr)](https://cran.r-project.org/package=qiitr)

## About Qiita

[Qiita](http://qiita.com/about) is a technical knowledge sharing and collaboration platform for programmers. API Document is here http://qiita.com/api/v2/docs.

## Installation

qiitr is now on CRAN.

```r
install.packages("qiitr")
```

If you need the development version, try:

```r
devtools::install_github("yutannihilation/qiitr")
```

## Preperation

Most of Qiita APIs need authorization by an access token. You can issue access tokens on [application settings](https://qiita.com/settings/applications) with the proper priviledge scope.

qiitr package uses `QIITA_ACCESSTOKEN` environmental variable for authoriation. Add the following to your `.Renviron` file. If you are not familiar with `.Renviron`, please read `?Startup`. The file is usually placed in the directory of `Sys.getenv("R_USER")`.

```r
QIITA_ACCESSTOKEN='(your access token)'
```

Or, you can temporarily set `QIITA_ACCESSTOKEN` by `qiita_set_accesstoken()`.

```r
qiita_set_accesstoken()
```

## Usage

### Get item information

Items means articles on Qiita. `qiita_get_items()` can get items by item IDs, tag IDs or user IDs.

```r
# get items by item ID
qiita_get_items(item_id = "7a78d897810446dd6a3b")

# get items by tag ID
qiita_get_items(tag_id = c("dplyr", "tidyr"), per_pages = 10L, page_limit = 1L)

# get items by user ID
qiita_get_items(user_id = "yutannihilation")
```

### Get user information

`qiita_get_authenticated_user()` returns the current user's information. `qiita_get_users()` returns the information about the specified user.

```r
# get the current user
qiita_get_authenticated_user()

# get a user by id
qiita_get_users("yutannihilation")
```

### Follow/Unfollow tags and users

You can also follow/unfollow  tags and users by qiitr functions. Note that thsese APIs requires write priviledge.

```r
# follow a user
qiita_follow_user("user1")

# unfollow a user
qiita_unfollow_user("user1")

# follow a tag
qiita_follow_tag("RStudio")

# unfollow a tag
qiita_unfollow_tag("RStudio")
```

### Post and edit items

`qiita_post_item()` posts an item. Note that the item is private by default.
You should manually check if the post is valid before make it public. `qiita_update_item()` updates it and `qiita_delete_item()` deletes it.

```r
# post an item
item <- qiita_post_item(title = "test", body = "This is an example.")

browseURL(item$url)

# update the item
qiita_update_item(item$id, title = "test", body = "**This is a strong example!**")

# delete the item
qiita_delete_item(item$id)
```

## Note

### Qiita:Team

Though this package doesn't provide full support for [Qiita:Team](https://teams.qiita.com/)-related APIs, you can set `QIITA_URL` environmental variable to change API endpoints. Please add the following to your `.Renviron`.

```r
QIITA_URL='(your Qiita:Team's URL)'
```


NEWS.md
# qiitr (development version)

# qiitr 0.1.1

* Fix a compatibility problem with R 4.0.

# qiitr 0.1.0

* First release on CRAN.


tests/testthat/test-apitest.R
context("util test")

test_that("qiita_tag test", {
  expect_equal(qiita_tag(name = "R", version = ">3.1"), list(name = "R", version = ">3.1"))
})


tests/testthat/test-qiita-tag.R
context("qiita_util_tag and qiita_util_payload test")

skip_on_cran()

test_that("qiita_util_tag test", {
  expect_equal(
    qiita_util_tag(name = "R", versions = ">3.1"),
    structure(list(name = "R", versions = I(">3.1")), class = c("qiita_tag", "list"))
  )
})

test_that("qiita_util_tag test without versions", {
  expect_equal(
    qiita_util_tag(name = "R"),
    structure(list(name = "R", versions = I(list())), class = c("qiita_tag", "list"))
  )
})

test_that("qiita_util_tag with multiple versions", {
  expect_equal(
    qiita_util_tag(name = "R", versions = c("3.1", "3.2")),
    structure(list(name = "R", versions = I(c("3.1", "3.2"))), class = c("qiita_tag", "list"))
  )
})



test_that("qiita_util_payload", {
  example_json <- '
{
  "body": "# Example",
  "title": "Example title",
  "tags": [
    {
      "name": "Ruby",
      "versions": [
        "0.0.1"
      ]
    }
  ],
  "private": false,
  "coediting": false
}'

  qiitr_json <- qiita_util_payload(body = "# Example",
                              coediting = FALSE,
                              private = FALSE,
                              tags = qiita_util_tag("Ruby", "0.0.1"),
                              title = "Example title")

  expect_equal(jsonlite::minify(qiitr_json),
               jsonlite::minify(example_json))
})

test_that("qiita_util_payload with multiple versions", {
  example_json <- '
{
  "body": "# Example",
  "title": "Example title",
  "tags": [
    {
      "name": "Ruby",
      "versions": [
        "0.0.1", "0.0.2"
      ]
    }
  ],
  "private": false,
  "coediting": false
}'

  qiitr_json <- qiita_util_payload(body = "# Example",
                              coediting = FALSE,
                              private = FALSE,
                              tags = qiita_util_tag("Ruby", c("0.0.1", "0.0.2")),
                              title = "Example title")

  expect_equal(jsonlite::minify(qiitr_json),
               jsonlite::minify(example_json))
})


tests/testthat/test-qiita-util.R
context("utilities")

skip_on_cran()

test_that("qiita_util_tag test", {
  expect_equal(
    qiita_util_tag(name = "R", versions = ">3.1"),
    structure(list(name = "R", versions = I(">3.1")), class = c("qiita_tag", "list"))
  )
})

test_that("qiita_util_tag test without versions", {
  expect_equal(
    qiita_util_tag(name = "R"),
    structure(list(name = "R", versions = I(list())), class = c("qiita_tag", "list"))
  )
})

test_that("qiita_util_tag with multiple versions", {
  expect_equal(
    qiita_util_tag(name = "R", versions = c("3.1", "3.2")),
    structure(list(name = "R", versions = I(c("3.1", "3.2"))), class = c("qiita_tag", "list"))
  )
})



test_that("qiita_util_payload", {
  example_json <- '
{
  "body": "# Example",
  "title": "Example title",
  "tags": [
    {
      "name": "Ruby",
      "versions": [
        "0.0.1"
      ]
    }
  ],
  "private": false,
  "coediting": false
}'

  qiitr_json <- qiita_util_payload(body = "# Example",
                              coediting = FALSE,
                              private = FALSE,
                              tags = qiita_util_tag("Ruby", "0.0.1"),
                              title = "Example title")

  expect_equal(jsonlite::minify(qiitr_json),
               jsonlite::minify(example_json))
})

test_that("qiita_util_payload with multiple versions", {
  example_json <- '
{
  "body": "# Example",
  "title": "Example title",
  "tags": [
    {
      "name": "Ruby",
      "versions": [
        "0.0.1", "0.0.2"
      ]
    }
  ],
  "private": false,
  "coediting": false
}'

  qiitr_json <- qiita_util_payload(body = "# Example",
                              coediting = FALSE,
                              private = FALSE,
                              tags = qiita_util_tag("Ruby", c("0.0.1", "0.0.2")),
                              title = "Example title")

  expect_equal(jsonlite::minify(qiitr_json),
               jsonlite::minify(example_json))
})


R/tag.r
#' Qiita Tag API
#'
#' Get, follow or unfollow tags via Qiita API.
#'
#' @name qiita_tag
#' @param tag_id Tag ID (e.g. \code{"R"}, \code{"dplyr"}).
#' @param user_id User ID (e.g. \code{"yutannihilation"}).
#' @param per_page Number of items per one page.
#' @param page_offset Page offset.
#' @param page_limit Max number of pages to aquire.
#' @examples
#' \dontrun{
#' # get a tag by Tag IDs
#' qiita_get_tags(tag_id = "R")
#'
#' # get tags by user ID
#' qiita_get_tags(user_id = "yutannihilation")
#'
#' # follow a tag
#' qiita_follow_tag("RStudio")
#'
#' # unfollow a tag
#' qiita_unfollow_tag("RStudio")
#' }
#' @export
qiita_get_tags <- function(tag_id = NULL, user_id = NULL,
                          per_page = 100L, page_offset = 0L, page_limit = 1L) {
  if(!is.null(tag_id) && !is.null(user_id))
    stop("You cannot specify tag_id and user_id both")

  if(is.null(tag_id) && is.null(user_id)) {
    path <- "/api/v2/tags"
    result <- qiita_api("GET", path = path,
                        per_page = per_page, page_offset = page_offset, page_limit = page_limit)
    return(result)
  }

  # Get an tag by ID (No pagenation is needed)
  if(!is.null(tag_id)){
    result <- purrr::map(tag_id, qiita_get_single_tag_by_id)
    return(result)
  }

  if(!is.null(user_id)){
    result <- qiita_get_tags_followed_by_user(user_id,
                                      per_page = per_page, page_offset = page_offset, page_limit = page_limit)
    return(result)
  }
}

qiita_get_single_tag_by_id <- function(tag_id) {
  path <- sprintf("/api/v2/tags/%s", tag_id)
  qiita_api("GET", path = path)
}

qiita_get_tags_followed_by_user <- function(user_id,
                                    per_page = 100L, page_offset = 0L, page_limit = 1L) {
  path <- sprintf("/api/v2/users/%s/following_tags", user_id)
  qiita_api("GET", path = path,
            per_page = per_page, page_offset = page_offset, page_limit = page_limit)
}

#' @rdname qiita_tag
#' @export
qiita_follow_tag <- function(tag_id) {
  if(!purrr::is_scalar_character(tag_id)) stop("tag_id must be a scalar character!")

  path <- sprintf("/api/v2/tags/%s/following", tag_id)
  qiita_api("PUT", path = path)
}


#' @rdname qiita_tag
#' @export
qiita_unfollow_tag <- function(tag_id) {
  if(!purrr::is_scalar_character(tag_id)) stop("tag_id must be a scalar character!")

  path <- sprintf("/api/v2/tags/%s/following", tag_id)
  qiita_api("DELETE", path = path)
}

#' @rdname qiita_tag
#' @export
qiita_is_following_tag <- function(tag_id) {
  if(!purrr::is_scalar_character(tag_id)) stop("tag_id must be a scalar character!")

 path <- sprintf("/api/v2/tags/%s/following", tag_id)
 qiita_api("GET", path = path, .expect204 = TRUE)
}


R/auth.r
#' Authorization with Qiita
#'
#' qiitr uses \code{QIITA_ACCESSTOKEN} environmental variable for authoriation. You can issue
#' a personal access token at \url{https://qiita.com/settings/applications}.
#' To set the envvar permanently, write \code{QIITA_ACCESSTOKEN = (your access token)} to a file
#' and save it as \code{.Renviron} on current directory or the home directory (For more info,
#' see \link[base]{Startup}). To set the variable temporarily, use \code{qiita_set_accesstoken}.
#'
#' @name qiita_auth
#' @export
qiita_set_accesstoken <- function() {
  if(!identical(Sys.getenv("QIITA_ACCESSTOKEN"), "")) {
    warning("QIITA_ACCESSTOKEN envvar is already set. Abort.")
    return(FALSE)
  }

  if(!interactive()) stop("Use Sys.setenv or .Renviron in noninteractive session.")

  msg <- "Personal access token for Qiita:"
  if(rstudioapi::isAvailable()) {
    token <- rstudioapi::askForPassword(msg)
  } else {
    cat(msg)
    token <- readLines(n = 1)
  }

  if(is.null(token) || identical(token, "")) {
    warning("Input was empty. Abort.")
    return(FALSE)
  }

  Sys.setenv(QIITA_ACCESSTOKEN = token)
  message(
    sprintf("To set QIITA_ACCESSTOKEN permanently, write the following line to a file and save it
as '.Renviron' on current directory or Sys.getenv('R_USER'):
QIITA_ACCESSTOKEN='%s'", token)
  )
}


R/item.r
#' Qiita Item (Article) API
#'
#' Get, post, delete, stock or unstock articles via Qiita API.
#'
#' @name qiita_item
#'
#' @param item_id Item (Article) ID.
#' @param query Query string (e.g. \code{"dplyr user:yutannihlation"}).
#' @param user_id User ID (e.g. \code{"yutannihilation"}).
#' @param per_page Number of items per one page.
#' @param page_offset Number of offset pages.
#' @param page_limit Max number of pages to retrieve.
#' @param tag_id Tag IDs (e.g. \code{"R"}).
#' @param body Content body.
#' @param title Title.
#' @param tags Tags. Use \code{qiita_util_tag} to generate tag objects.
#' @param private If \code{TRUE}, the post will be private.
#' @param coediting If \code{TRUE}, the post will be editable by team members.
#' @param gist If \code{TRUE}, post the code to Gist.
#' @param tweet If \code{TRUE}, notify on Twitter.
#'
#' @examples
#' \dontrun{
#' # get items by item ID
#' qiita_get_items(item_id = "7a78d897810446dd6a3b")
#'
#' # get items by tag ID
#' qiita_get_items(tag_id = c("dplyr", "tidyr"), per_pages = 10L, page_limit = 1L)
#'
#' # get items by user ID
#' qiita_get_items(user_id = "yutannihilation")
#'
#' # Post an item. Note that the post is private by default.
#' # You should manually check if the post is valid before make it public.
#' item <- qiita_post_item(title = "test", body = "This is an example.")
#'
#' # update the post
#' qiita_update_item(item$id, title = "test", body = "**This is a strong example!**")
#'
#' # delete the post
#' qiita_delete_item(item$id)
#' }
#'
#' @export
qiita_get_items <- function (item_id = NULL, tag_id = NULL, user_id = NULL, query = NULL,
                            per_page = 100L, page_offset = 0L, page_limit = 1L) {
  num_of_not_nulls <- sum(!purrr::map_lgl(list(item_id, tag_id, user_id), is.null))

  if (num_of_not_nulls > 1)
    stop("You cannot specify more-than-one conditions")

  # Get newest items
  if (num_of_not_nulls == 0) {
    path <- "/api/v2/items"
    result <- qiita_api("GET", path = path,
                        query = list(query = query),
                        per_page = per_page, page_offset = page_offset, page_limit = page_limit)
    return(result)
  }

  # Get items by ID (No pagenation is needed)
  if (!is.null(item_id)) {
    result <- purrr::map(item_id, qiita_get_single_item)
    return(result)
  }

  # Get items by tag
  if (!is.null(tag_id)) {
    result <- purrr::map(tag_id, qiita_get_items_by_tag,
                         per_page = per_page, page_offset = page_offset, page_limit = page_limit)
    return(purrr::flatten(result))
  }

  # Get items by user
  if (!is.null(user_id)) {
    result <- purrr::map(user_id, qiita_get_items_by_user,
                         per_page = per_page, page_offset = page_offset, page_limit = page_limit)
    return(purrr::flatten(result))

  }
}

qiita_get_single_item <- function(item_id) {
  path <- sprintf("/api/v2/items/%s", item_id)
  qiita_api("GET", path = path)
}

qiita_get_items_by_tag <- function(tag_id, per_page, page_offset, page_limit) {
  path <- sprintf("/api/v2/tags/%s/items", tag_id)
  qiita_api("GET", path = path,
            per_page = per_page, page_offset = page_offset, page_limit = page_limit)
}

qiita_get_items_by_user <- function(user_id, per_page, page_offset, page_limit) {
  path <- sprintf("/api/v2/users/%s/items", user_id)
  qiita_api("GET", path = path,
            per_page = per_page, page_offset = page_offset, page_limit = page_limit)
}

#' @rdname qiita_item
#' @export
qiita_post_item <- function(title, body, tags = qiita_util_tag("R"),
                            coediting = FALSE, private = TRUE, gist = FALSE, tweet = FALSE) {
  if(!purrr::is_scalar_character(title)) stop("title must be a scalar character!")
  if(!purrr::is_scalar_character(body)) stop("body must be a scalar character!")

  path    <- "/api/v2/items"
  payload <- qiita_util_payload(
    body = body,
    title = title,
    tags = tags,
    private = private,
    coediting = coediting,
    gist = gist,
    tweet = tweet
  )

  qiita_api("POST", path = path,
            payload = payload)
}

#' @rdname qiita_item
#' @export
qiita_delete_item <- function(item_id) {
  if(!purrr::is_scalar_character(item_id)) stop("item_id must be a scalar character!")

  path <- sprintf("/api/v2/items/%s", item_id)
  qiita_api("DELETE", path = path)
}

#' @rdname qiita_item
#' @export
qiita_update_item <- function(item_id, title, body,
                              tags = list(qiita_util_tag("R")), private = TRUE) {
  if(!purrr::is_scalar_character(item_id)) stop("item_id must be a scalar character!")
  if(!purrr::is_scalar_character(title)) stop("title must be a scalar character!")
  if(!purrr::is_scalar_character(body)) stop("body must be a scalar character!")

  path <- sprintf("/api/v2/items/%s", item_id)
  payload <- qiita_util_payload(
    body = body,
    title = title,
    tags = tags,
    private = private
  )

  qiita_api("PATCH", path = path,
            payload = payload)
}

#' @rdname qiita_item
#' @export
qiita_stock_item <- function(item_id) {
  if(!purrr::is_scalar_character(item_id)) stop("item_id must be a scalar character!")

  path <- sprintf("/api/v2/items/%s/stock", item_id)
  qiita_api("PUT", path = path)
}

#' @rdname qiita_item
#' @export
qiita_unstock_item <- function(item_id) {
  if(!purrr::is_scalar_character(item_id)) stop("item_id must be a scalar character!")

  path <- sprintf("/api/v2/items/%s/stock", item_id)
  qiita_api("DELETE", path = path)
}

#' @rdname qiita_item
#' @export
qiita_is_stocked_item <- function(item_id) {
  if(!purrr::is_scalar_character(item_id)) stop("item_id must be a scalar character!")

  path <- sprintf("/api/v2/items/%s/stock", item_id)
  result <- qiita_api("GET", path = path, .expect204 = TRUE)
}

#' @rdname qiita_item
#' @export
qiita_get_stocks <- function(user_id,
                             per_page = 100L, page_offset = 0L, page_limit = 1L) {

  result <- purrr::map(user_id, qiita_get_stocks_by_single_user,
                       per_page = per_page, page_offset = page_offset, page_limit = page_limit)
  return(result)
}

qiita_get_stocks_by_single_user <- function(user_id, per_page, page_offset, page_limit) {
  path <- sprintf("/api/v2/users/%s/stocks", user_id)
  qiita_api("GET", path = path,
            per_page = per_page, page_offset = page_offset, page_limit = page_limit)
}


R/user.r
#' Qiita User API
#'
#' Get, follow or unfollow users via Qiita API.
#'
#' @name qiita_user
#' @param user_id User ID (e.g. \code{"yutannihilation"}).
#' @param item_id Item (article) ID.
#' @param per_page Number of items per one page.
#' @param page_offset Number of offset pages.
#' @param page_limit Max number of pages to retrieve.
#'
#' @examples
#' \dontrun{
#' # get a user by id
#' qiita_get_users("yutannihilation")
#'
#' # follow a user
#' qiita_follow_user("user1")
#'
#' # unfollow a user
#' qiita_unfollow_user("user1")
#'
#' # get the current user
#' qiita_get_authenticated_user()
#' }
#'
#' @export
qiita_get_stockers <- function(item_id,
                              per_page = 100L, page_offset = 0L, page_limit = 1L) {
  result <- purrr::map(item_id, qiita_get_stockers_by_single_item,
             per_page = per_page, page_offset = page_offset, page_limit = page_limit)
  return(purrr::flatten(result))
}

qiita_get_stockers_by_single_item <- function(item_id, per_page, page_offset, page_limit) {
  path <- sprintf("/api/v2/items/%s/stockers", item_id)
  qiita_api("GET", path = path,
            per_page = per_page, page_offset = page_offset, page_limit = page_limit)
}

#' @rdname qiita_user
#' @export
qiita_get_users <- function(user_id) {
  purrr::map(user_id, qiita_get_single_user)
}

qiita_get_single_user <- function(user_id) {
  path <- sprintf("/api/v2/users/%s", user_id)
  qiita_api("GET", path = path)
}

#' @rdname qiita_user
#' @export
qiita_get_followees <- function(user_id,
                               per_page = 100L, page_offset = 0L, page_limit = 1L) {
  result <- purrr::map(user_id, qiita_get_followees_by_single_user,
                       per_page = per_page, page_offset = page_offset, page_limit = page_limit)
  return(purrr::flatten(result))
}

qiita_get_followees_by_single_user <- function(user_id, per_page, page_offset, page_limit) {
  path <- sprintf("/api/v2/users/%s/followees", user_id)
  qiita_api("GET", path = path,
            per_page = per_page, page_offset = page_offset, page_limit = page_limit)
}

#' @rdname qiita_user
#' @export
qiita_get_followers <- function(user_id,
                               per_page = 100L, page_offset = 0L, page_limit = 1L) {
  result <- purrr::map(user_id, qiita_get_followers_by_single_user,
                       per_page = per_page, page_offset = page_offset, page_limit = page_limit)
  return(purrr::flatten(result))
}

qiita_get_followers_by_single_user <- function(user_id, per_page, page_offset, page_limit) {
  path <- sprintf("/api/v2/users/%s/followers", user_id)
  qiita_api("GET", path = path,
            per_page = per_page, page_offset = page_offset, page_limit = page_limit)
}

#' @rdname qiita_user
#' @export
qiita_follow_user <- function(user_id) {
  if(!purrr::is_scalar_character(user_id)) stop("user_id must be a scalar character!")

  path <- sprintf("/api/v2/users/%s/following", user_id)
  qiita_api("PUT", path = path)
}

#' @rdname qiita_user
#' @export
qiita_unfollow_user <- function(user_id) {
  if(!purrr::is_scalar_character(user_id)) stop("user_id must be a scalar character!")

  path <- sprintf("/api/v2/users/%s/following", user_id)
  qiita_api("DELETE", path = path)
}

#' @rdname qiita_user
#' @export
qiita_is_following_user <- function(user_id) {
  if(!purrr::is_scalar_character(user_id)) stop("user_id must be a scalar character!")

  path <- sprintf("/api/v2/users/%s/following", user_id)
  qiita_api("GET", path = path, .expect204 = TRUE)
}

#' @rdname qiita_user
#' @export
qiita_get_authenticated_user <- function() {
  path <- "/api/v2/authenticated_user"
  qiita_api("GET", path = path)
}


R/Qiita.r
#' An R6 Class for Qiita API
#'
#' @docType class
#' @import R6
#' @export
Qiita <- R6::R6Class(
  "Qiita",
  public = list(
    token = NA,
    url   = NA,

    initialize = function(token, url = "https://qiita.com") {
      self$token <- token
      self$url   <- url
    },

    get_tag           = function(tag_id = NULL, user_id = NULL, per_page = 100L, page_offset = 0L, page_limit = 1L)
      qiita_get_tag(url = self$url, token = self$token, tag_id = tag_id, user_id = user_id,
                    per_page = per_page, page_offset = page_offset, page_limit = page_limit),
    get_following_tag = function(user_id, per_page = 100L, page_offset = 0L, page_limit = 1L)
      qiita_get_following_tag(url = self$url, token = self$token, user_id = user_id,
                              per_page = per_page, page_offset = page_offset, page_limit = page_limit),
    delete_tagfollow  = function(tag_id) qiita_delete_tagfollow(url = self$url, token = self$token, tag_id = tag_id),
    get_tagfollow     = function(tag_id, per_page = 100L, page_offset = 0L, page_limit = 1L)
      qiita_get_tagfollow(url = self$url, token = self$token, tag_id = tag_id,
                          per_page = per_page, page_offset = page_offset, page_limit = page_limit),
    put_tagfollow     = function(tag_id) qiita_put_tagfollow(url = self$url, token = self$token, tag_id = tag_id),
    get_stocker       = function(item_id, per_page = 100L, page_offset = 0L, page_limit = 1L)
      qiita_get_stocker(url = self$url, token = self$token, item_id = item_id,
                        per_page = per_page, page_offset = page_offset, page_limit = page_limit),
    get_user          = function(user_id) qiita_get_user(url = self$url, token = self$token, user_id = user_id),
    get_followee      = function(user_id, per_page = 100L, page_offset = 0L, page_limit = 1L)
      qiita_get_followee(url = self$url, token = self$token, user_id = user_id,
                         per_page = per_page, page_offset = page_offset, page_limit = page_limit),
    get_follower      = function(user_id, per_page = 100L, page_offset = 0L, page_limit = 1L)
      qiita_get_follower(url = self$url, token = self$token, user_id = user_id,
                         per_page = per_page, page_offset = page_offset, page_limit = page_limit),
    delete_userfollow = function(user_id) qiita_delete_userfollow(url = self$url, token = self$token, user_id = user_id),
    get_userfollow    = function(user_id, per_page = 100L, page_offset = 0L, page_limit = 1L)
      qiita_get_userfollow(url = self$url, token = self$token, user_id = user_id,
                           per_page = per_page, page_offset = page_offset, page_limit = page_limit),
    put_userfollow    = function(user_id) qiita_put_userfollow(url = self$url, token = self$token, user_id = user_id),
    get_authenticated_user = function() qiita_get_authenticated_user(url = self$url, token = self$token),
    get_comment       = function(comment_id = NULL, item_id = NULL, per_page = 100L, page_offset = 0L, page_limit = 1L)
      qiita_get_comment(url = self$url, token = self$token, comment_id = comment_id, item_id = item_id,
                        per_page = per_page, page_offset = page_offset, page_limit = page_limit),
    delete_comment    = function(comment_id) qiita_delete_comment(url = self$url, token = self$token, comment_id = commend_id),
    patch_comment     = function(commend_id, body) qiita_patch_comment(url = self$url, token = self$token, commend_id = commend_id, body = body),
    post_comment      = function(item_id, body) qiita_post_comment(url = self$url, token = self$token, item_id = item_id, body = body),
    get_item          = function(item_id = NULL, tag_id = NULL, user_id = NULL, query = NULL, per_page = 100L, page_offset = 0L, page_limit = 1L)
      qiita_get_item(url = self$url, token = self$token, item_id = item_id, tag_id = tag_id, user_id = user_id, query = query,
                     per_page = per_page, page_offset = page_offset, page_limit = page_limit),
    post_item         = function(title, body, tags = list(qiita_tag("R")), gist = FALSE, private = FALSE, tweet = FALSE) qiita_post_item(url = self$url, token = self$token, title = title, body = body, tags = tags, gist = gist, private = private, tweet = tweet),
    delete_item       = function(item_id) qiita_delete_item(url = self$url, token = self$token, item_id = item_id),
    patch_item        = function(item_id, title, body, tags = list(qiita_tag("R")), private = FALSE) qiita_patch_item(url = self$url, token = self$token, item_id = item_id, title = title, body = body, tags = tags, private = private),
    delete_stock      = function(item_id) qiita_delete_stock(url = self$url, token = self$token, item_id = item_id),
    get_stock         = function(item_id = NULL, user_id = NULL, per_page = 100L, page_offset = 0L, page_limit = 1L)
      qiita_get_stock(url = self$url, token = self$token, item_id = item_id, user_id = user_id,
                      per_page = per_page, page_offset = page_offset, page_limit = page_limit),
    put_stock         = function(item_id) qiita_put_stock(url = self$url, token = self$token, item_id = item_id)
  )
)


R/comment.r
#' Qiita Comments API
#'
#' Get, write, update or delete comments via Qiita API.
#'
#' @name qiita_comment
#' @param comment_id Comment ID.
#' @param item_id Item (article) ID.
#' @param per_page Number of items per one page.
#' @param page_offset Number of offset pages.
#' @param page_limit Max number of pages to retrieve.
#' @param body body of the item
#' @examples
#' \dontrun{
#' # get a comment by id
#' qiita_get_comments(comment_id = "1fdbb164e19d79e10203")
#'
#' # get comments by item id
#' qiita_get_comments(item_id = "b4130186e1e095719dcb")
#'
#' # post a comment to some item
#' qiita_post_comment(item_id = "123456789", body = "Thank you!!!")
#' }
#' @export
qiita_get_comments <- function(comment_id = NULL, item_id = NULL,
                              per_page = 100L, page_offset = 0L, page_limit = 1L) {
  if(!is.null(comment_id) && !is.null(item_id)) stop("You cannot specify comment_id and item_id both")
  if(is.null(comment_id) && is.null(item_id))   stop("Please specify commend_id or item_id")

  # Get a comment by ID (No pagenation is needed)
  if(!is.null(comment_id)){
    result <- purrr::map(comment_id, qiita_get_single_comment_by_id)
    return(result)
  }

  if(!is.null(item_id)) {
    result <- purrr::map(item_id, qiita_get_comments_by_item,
                         per_page = per_page, page_offset = page_offset, page_limit = page_limit)
    return(purrr::flatten(result))
  }
}

qiita_get_single_comment_by_id <- function(comment_id) {
  path <- sprintf("/api/v2/comments/%s", comment_id)
  qiita_api("GET", path = path)
}

qiita_get_comments_by_item <- function(item_id, per_page, page_offset, page_limit) {
  path <- sprintf("/api/v2/items/%s/comments", item_id)
  result <- qiita_api("GET", path = path,
                      per_page = per_page, page_offset = page_offset, page_limit = page_limit)
}

#' @rdname qiita_comment
#' @export
qiita_delete_comment <- function(comment_id) {
  if(!purrr::is_scalar_character(comment_id)) stop("comment_id must be a scalar character!")

  path <- sprintf("/api/v2/comments/%s", comment_id)
  qiita_api("DELETE", path = path)
}

#' @rdname qiita_comment
#' @export
qiita_update_comment <- function(comment_id, body) {
  if(!purrr::is_scalar_character(comment_id)) stop("comment_id must be a scalar character!")
  if(!purrr::is_scalar_character(body)) stop("body must be a scalar character!")

  path <- sprintf("/api/v2/comments/%s", comment_id)
  qiita_api("PATCH", path = path,
            payload = qiita_util_payload(body = body))
}

#' @rdname qiita_comment
#' @export
qiita_post_comment <- function(item_id, body) {
  if(!purrr::is_scalar_character(item_id)) stop("item_id must be a scalar character!")
  if(!purrr::is_scalar_character(body)) stop("body must be a scalar character!")

  path <- sprintf("/api/v2/items/%s/comments", item_id)
  qiita_api("POST", path = path,
            payload = qiita_util_payload(body = body))
}


R/util.r
#' Generate Payload And Tag For Qiita API
#'
#' @name qiita_util
#'
#' @param name Tag name
#' @param versions Versions (e.g. \code{3.1}, \code{>3.2}).
#' @param body Content body.
#' @param title Title.
#' @param tags Tags. Use \code{qiita_util_tag} to generate tag objects.
#' @param private If \code{TRUE}, the post will be private.
#' @param coediting If \code{TRUE}, the post will be editable by team members.
#' @param gist If \code{TRUE}, post the code to Gist.
#' @param tweet If \code{TRUE}, notify on Twitter.
#'
#' @examples
#' qiita_util_tag(name = "R", versions = ">3.1")
#'
#' qiita_util_payload(body = "foo",
#'                    title = "test",
#'                    tags = list(
#'                      qiita_util_tag(name = "R", versions = ">3.1"),
#'                      qiita_util_tag(name = "dplyr")
#'                    ),
#'                    private = TRUE)
#'
#' @export
qiita_util_tag <- function(name, versions = NULL) {
  x <- list()
  class(x) <- c("qiita_tag", "list")

  x$name     <- name
  x$versions <- if(is.null(versions)) I(list()) else I(versions)
  x
}

#' @rdname qiita_util
#' @export
qiita_util_payload <- function(body = NULL, title = NULL, tags = NULL,
                          private = NULL, coediting = NULL,
                          gist = NULL, tweet = NULL) {
  # tag must be wrapped with list.
  if(inherits(tags, "qiita_tag")) tags <- list(tags)

  payload <- list()

  payload$body   <- body
  payload$title  <- title
  payload$tags   <- tags
  payload$private <- private
  payload$coediting <- coediting
  payload$gist   <- gist
  payload$tweet  <- tweet

  jsonlite::toJSON(payload, auto_unbox = TRUE)
}



#' Send A Request To Qiita API
#'
#' An \link[httr]{httr} Wrapper for Qiita API.
#'
#' @name qiita_api
#' @import httr
#' @param verb Method type (e.g. GET, POST, DELETE).
#' @param path Path of API.
#' @param payload JSON payload.
#' @param query Query strings.
#' @param per_page Number of entries (e.g. items, tags, users) in one page.
#' @param page_offset Number of offset pages.
#' @param page_limit Max number of pages to retrieve.
#' @param .expect204
#'   If \code{TRUE}, return \code{TRUE} for status code 204 and \code{FALSE} for status code 404.
#' @export
qiita_api <- function(verb, path, payload = NULL, query = NULL,
                      per_page = 100L, page_offset = 0L, page_limit = 1L,
                      .expect204 = FALSE) {

  # Set QIITA_URL envvar to access Qiita API.
  url <- Sys.getenv("QIITA_URL")
  if(identical(url, "")) url <- "https://qiita.com/"

  # Set QIITA_ACCESSTOKEN envvar before.
  token <- Sys.getenv("QIITA_ACCESSTOKEN")
  if(identical(token, "")) {
    warning("Please set QIITA_ACCESSTOKEN envvar to use APIs that needs authorization.")
  }

  if (is.null(query)) query <- list()
  query$per_page <- per_page
  query$page     <- start_page <- 1 + page_offset

  header <- httr::add_headers(`Content-Type`  = "application/json",
                              `Authorization` = paste("Bearer", token))

  res <- httr::VERB(
    verb, url = url, path = path,
    config = header, body = payload, query = query
  )

  # for qiita_is_following_{user,tag}
  if (.expect204) {
    status_code <- httr::status_code(res)
    if(identical(status_code, 204L)) return(TRUE)
    if(identical(status_code, 404L)) return(FALSE)
    stop(sprintf("Unknown status code %d: ", status_code))
  }

  # FIXME: specifying `type` is a temporal workaround to avoid error with mime::guess_type()
  first_result <- qiia_api_content(res)

  total_count <- as.integer(res$headers$`total-count`)
  total_page  <- ceiling(total_count / per_page)
  end_page    <- as.integer(min(total_page, page_offset + page_limit))

  message(sprintf("total count is %d (= %d pages)", total_count, total_page))

  if (end_page <= start_page) {
    return(first_result)
  }

  result <- vector("list", end_page - start_page + 1)
  result[[1]] <- first_result

  # start with page 2 as we already have page 1.
  for (i in seq(2, length(result))) {
    p <- i + start_page - 1
    message(sprintf("getting page %s...", p))
    query$page <- p

    res <- httr::VERB(
      verb, url = url, path = path,
      config = header, body = payload, query = query
    )

    result[[i]] <- qiia_api_content(res)
  }

  purrr::flatten(result)
}

qiia_api_content <- function(res) {
  status_code <- httr::status_code(res)

  if (status_code >= 400) {
    stop(sprintf("API returns an error (HTTP %d):\n    %s",
                 status_code,  httr::content(res, encoding = "UTF-8", as = "text")))
  }

  httr::content(res, encoding = "UTF-8", type = "application/json")
}
